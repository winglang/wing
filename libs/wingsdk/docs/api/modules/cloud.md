[@monadahq/wingsdk](../README.md) / [Exports](../modules.md) / cloud

# Namespace: cloud

## Table of contents

### Enumerations

- [BucketInflightMethods](../enums/cloud.BucketInflightMethods.md)
- [FunctionInflightMethods](../enums/cloud.FunctionInflightMethods.md)
- [QueueInflightMethods](../enums/cloud.QueueInflightMethods.md)

### Classes

- [Bucket](../classes/cloud.Bucket.md)
- [BucketBase](../classes/cloud.BucketBase.md)
- [Function](../classes/cloud.Function.md)
- [FunctionBase](../classes/cloud.FunctionBase.md)
- [Queue](../classes/cloud.Queue.md)
- [QueueBase](../classes/cloud.QueueBase.md)
- [Resource](../classes/cloud.Resource.md)

### Interfaces

- [BucketProps](../interfaces/cloud.BucketProps.md)
- [FunctionProps](../interfaces/cloud.FunctionProps.md)
- [IBucketClient](../interfaces/cloud.IBucketClient.md)
- [IFunctionClient](../interfaces/cloud.IFunctionClient.md)
- [IQueueClient](../interfaces/cloud.IQueueClient.md)
- [QueueOnMessageProps](../interfaces/cloud.QueueOnMessageProps.md)
- [QueueProps](../interfaces/cloud.QueueProps.md)

### Variables

- [BUCKET\_TYPE](cloud.md#bucket_type)
- [FUNCTION\_TYPE](cloud.md#function_type)
- [QUEUE\_TYPE](cloud.md#queue_type)

## Variables

### BUCKET\_TYPE

• `Const` **BUCKET\_TYPE**: ``"wingsdk.cloud.Bucket"``

Global identifier for `Bucket`.

#### Defined in

[src/cloud/bucket.ts:9](https://github.com/monadahq/winglang/blob/438eedb/libs/wingsdk/src/cloud/bucket.ts#L9)

___

### FUNCTION\_TYPE

• `Const` **FUNCTION\_TYPE**: ``"wingsdk.cloud.Function"``

Global identifier for `Function`.

#### Defined in

[src/cloud/function.ts:9](https://github.com/monadahq/winglang/blob/438eedb/libs/wingsdk/src/cloud/function.ts#L9)

___

### QUEUE\_TYPE

• `Const` **QUEUE\_TYPE**: ``"wingsdk.cloud.Queue"``

Global identifier for `Queue`.

#### Defined in

[src/cloud/queue.ts:10](https://github.com/monadahq/winglang/blob/438eedb/libs/wingsdk/src/cloud/queue.ts#L10)
