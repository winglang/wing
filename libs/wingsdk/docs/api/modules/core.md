[@monadahq/wingsdk](../README.md) / [Exports](../modules.md) / core

# Namespace: core

## Table of contents

### Enumerations

- [Language](../enums/core.Language.md)

### Classes

- [App](../classes/core.App.md)
- [Code](../classes/core.Code.md)
- [DependencyGraph](../classes/core.DependencyGraph.md)
- [DependencyVertex](../classes/core.DependencyVertex.md)
- [Duration](../classes/core.Duration.md)
- [Inflight](../classes/core.Inflight.md)
- [InflightClient](../classes/core.InflightClient.md)
- [NodeJsCode](../classes/core.NodeJsCode.md)
- [Synthesizer](../classes/core.Synthesizer.md)
- [Testing](../classes/core.Testing.md)

### Interfaces

- [AppProps](../interfaces/core.AppProps.md)
- [Capture](../interfaces/core.Capture.md)
- [CaptureMetadata](../interfaces/core.CaptureMetadata.md)
- [ICapturable](../interfaces/core.ICapturable.md)
- [ICapturableConstruct](../interfaces/core.ICapturableConstruct.md)
- [InflightBundleOptions](../interfaces/core.InflightBundleOptions.md)
- [InflightProps](../interfaces/core.InflightProps.md)
- [SynthesizerProps](../interfaces/core.SynthesizerProps.md)
