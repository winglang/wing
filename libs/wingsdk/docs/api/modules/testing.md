[@monadahq/wingsdk](../README.md) / [Exports](../modules.md) / testing

# Namespace: testing

## Table of contents

### Classes

- [Simulator](../classes/testing.Simulator.md)

### Interfaces

- [IResourceResolver](../interfaces/testing.IResourceResolver.md)
- [ISimulatorDispatcher](../interfaces/testing.ISimulatorDispatcher.md)
- [SimulatorContext](../interfaces/testing.SimulatorContext.md)
- [SimulatorProps](../interfaces/testing.SimulatorProps.md)
