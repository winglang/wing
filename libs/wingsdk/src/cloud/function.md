---
title: Function
id: function
description: A built-in resource for creating serverless functions.
keywords:
  [
    Wing reference,
    Wing language,
    language,
    Wing standard library,
    Wing programming language,
    Serverless function,
  ]
sidebar_position: 1
---

The `cloud.Function` resource represents a serverless function for performing short, stateless tasks.
Functions are typically used to run business logic in response to events, such as a file being uploaded to a bucket, a message being pushed to a queue, or a timer expiring.

When a function is invoked on a cloud provider, it is typically executed in a container/host which is provisioned on demand.

Functions may be invoked more than once, and some cloud providers may automatically retry failed invocations.
For performance reasons, most cloud providers impose a timeout on functions, after which the function is automatically terminated.

## Usage

A function can be invoked in two ways:

* **invoke()** - Executes the function with a payload and waits for the result.
* **invokeAsync()** - Kicks off the execution of the function with a payload and returns immediately while the function is running.

```ts playground
bring cloud;
bring util;

// defining a cloud.Function resource
let countWords = new cloud.Function(inflight (s: str): str => {
  return "${s.split(" ").length}";
}) as "countWords";

let longTask = new cloud.Function(inflight () => {
  util.sleep(30s);
  log("done!");
});

new cloud.Function(inflight () => {
  let sentence = "I am a sentence with 7 words";
  // invoking cloud.Function from inflight context
  let wordsCount = countWords.invoke(sentence);
  log("'${sentence}' has ${wordsCount} words");

  longTask.invokeAsync("");
  log("task started");
}) as "Invoke Me";
```

## Function container reuse

Most cloud providers will opportunistically reuse the function's container in additional invocations. It is possible
to leverage this behavior to cache objects across function executions using `inflight new` and inflight fields.

The following example reads the `bigdata.json` file once and reuses it every time `query()` is called.

```js
bring cloud;

let big = new cloud.Bucket();

big.addObject("bigdata.json", Json.stringify({
  "my-key": "my-value"
}));

class MyDatabase {
  inflight bigdata: Json;
  inflight new() {
    // download big data once
    this.bigdata = big.getJson("bigdata.json");
  }

  pub inflight query(key: str): Json {
    return this.bigdata.get(key);
  }
}

let db = new MyDatabase();

new cloud.Function(inflight () => {
  log(Json.stringify(db.query("my-key")));
});```

## Target-specific details

### Simulator (`sim`)

The sim implementation of `cloud.Function` runs the inflight code as a JavaScript function.

### AWS (`tf-aws` and `awscdk`)

The AWS implementation of `cloud.Function` uses [AWS Lambda](https://aws.amazon.com/lambda/).

To add extra IAM permissions to the function, you can use the `aws.Function` class as shown below.

```ts playground
bring aws;
bring cloud;

let f = new cloud.Function(inflight () => {
  log("Hello world!");
});
if let lambdaFn = aws.Function.from(f) {
  lambdaFn.addPolicyStatements(
    aws.PolicyStatement {
      actions: ["ses:sendEmail"],
      effect: aws.Effect.ALLOW,
      resources: ["*"],
    },
  );
}
```

### Azure (`tf-azure`)

The Azure implementation of `cloud.Function` uses [Azure Functions](https://azure.microsoft.com/en-us/products/functions).

ðŸš§ `invoke` API is not supported yet (tracking issue: [#1371](https://github.com/winglang/wing/issues/1371))

### GCP (`tf-gcp`)

ðŸš§ Not supported yet (tracking issue: [#614](https://github.com/winglang/wing/issues/614))
