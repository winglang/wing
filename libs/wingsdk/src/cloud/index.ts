export * from "./api";
export * from "./bucket";
export * from "./counter";
export * from "./domain";
export * from "./endpoint";
export * from "./function";
export * from "./on-deploy";
export * from "./queue";
export * from "./schedule";
export * from "./secret";
export * from "./service";
export * from "./topic";
export * from "./website";
