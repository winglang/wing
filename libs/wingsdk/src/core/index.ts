export * from "./app";
export * from "./attributes";
export * from "./connections";
export * from "./dependency";
export * from "./inflight";
export * from "./lifting";
export * from "./tree";
