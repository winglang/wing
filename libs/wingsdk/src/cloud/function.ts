import { mkdirSync, writeFileSync } from "fs";
import { join } from "path";
import { Construct } from "constructs";
import { fqnForType } from "../constants";
import { App } from "../core";
import { CaseConventions, ResourceNames } from "../shared/resource-names";
import { Duration, IInflight, IInflightHost, Node, Resource } from "../std";

/**
 * Global identifier for `Function`.
 */
export const FUNCTION_FQN = fqnForType("cloud.Function");

/**
 * Options for `Function`.
 */
export interface FunctionProps {
  /**
   * Environment variables to pass to the function.
   * @default - No environment variables.
   */
  readonly env?: { [key: string]: string };

  /**
   * The maximum amount of time the function can run.
   * @default 1m
   */
  readonly timeout?: Duration;

  /**
   * The amount of memory to allocate to the function, in MB.
   * @default 1024
   */
  readonly memory?: number;

  /**
   * Specifies the number of days that function logs will be kept.
   * Setting negative value means logs will not expire.
   * @default 30
   */
  readonly logRetentionDays?: number;
}

/**
 * A function.
 *
 * @inflight `@winglang/sdk.cloud.IFunctionClient`
 * @abstract
 */
export class Function extends Resource implements IInflightHost {
  private readonly _env: Record<string, string> = {};

  /**
   * The path to the entrypoint source code of the function.
   */
  protected readonly entrypoint!: string;

  constructor(
    scope: Construct,
    id: string,
    handler: IFunctionHandler,
    props: FunctionProps = {}
  ) {
    if (new.target === Function) {
      return Resource._newFromFactory(FUNCTION_FQN, scope, id, handler, props);
    }

    super(scope, id);

    Node.of(this).title = "Function";
    Node.of(this).description = "A cloud function (FaaS)";

    for (const [key, value] of Object.entries(props.env ?? {})) {
      this.addEnvironment(key, value);
    }

    // indicates that we are calling the inflight constructor and the
    // inflight "handle" method on the handler resource.
    handler._registerOnLift(this, ["handle", "$inflight_init"]);

    const lines = this._getCodeLines(handler);
    const assetName = ResourceNames.generateName(this, {
      // Avoid characters that may cause path issues
      disallowedRegex: /[><:"/\\|?*\s]/g,
      case: CaseConventions.LOWERCASE,
      sep: "_",
    });

    // write the entrypoint next to the partial inflight code emitted by the compiler, so that
    // `require` resolves naturally.

    const workdir = App.of(this).workdir;
    mkdirSync(workdir, { recursive: true });
    const entrypoint = join(workdir, `${assetName}.js`);
    writeFileSync(entrypoint, lines.join("\n"));
    this.entrypoint = entrypoint;

    if (process.env.WING_TARGET) {
      this.addEnvironment("WING_TARGET", process.env.WING_TARGET);
    }
  }

  /**
   * @internal
   * @param handler IFunctionHandler
   * @returns the function code lines as strings
   */
  protected _getCodeLines(handler: IFunctionHandler): string[] {
    const inflightClient = handler._toInflight();
    const lines = new Array<string>();

    lines.push('"use strict";');
    lines.push("exports.handler = async function(event) {");
    lines.push(`  return await (${inflightClient}).handle(event);`);
    lines.push("};");

    return lines;
  }

  /**
   * Add an environment variable to the function.
   */
  public addEnvironment(name: string, value: string) {
    if (this._env[name] !== undefined && this._env[name] !== value) {
      throw new Error(
        `Environment variable "${name}" already set with a different value.`
      );
    }
    this._env[name] = value;
  }

  /**
   * Returns the set of environment variables for this function.
   */
  public get env(): Record<string, string> {
    return { ...this._env };
  }
}

/**
 * Inflight interface for `Function`.
 */
export interface IFunctionClient {
  /**
   * Invokes the function with a payload and waits for the result.
   * @inflight
   */
  invoke(payload: string): Promise<string>;

  /**
   * Kicks off the execution of the function with a payload and returns immediately while the function is running.
   * @inflight
   */
  invokeAsync(payload: string): Promise<void>;
}

/**
 * A resource with an inflight "handle" method that can be used to
 * create a `cloud.Function`.
 *
 * @inflight `@winglang/sdk.cloud.IFunctionHandlerClient`
 */
export interface IFunctionHandler extends IInflight {}

/**
 * Inflight client for `IFunctionHandler`.
 */
export interface IFunctionHandlerClient {
  /**
   * Entrypoint function that will be called when the cloud function is invoked.
   * @inflight
   */
  handle(event: string): Promise<string | undefined>;
}

/**
 * List of inflight operations available for `Function`.
 * @internal
 */
export enum FunctionInflightMethods {
  /** `Function.invoke` */
  INVOKE = "invoke",
  /** `Function.invokeAsync` */
  INVOKE_ASYNC = "invokeAsync",
}
