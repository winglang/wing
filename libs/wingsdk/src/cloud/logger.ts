import { Construct, IConstruct } from "constructs";
import { Polycons } from "polycons";
import { Code } from "../core/inflight";
import { Resource } from "../core/resource";

export const LOGGER_TYPE = "wingsdk.cloud.Logger";
export const LOGGER_SYMBOL = Symbol.for(LOGGER_TYPE);

/**
 * Functionality shared between all `Logger` implementations.
 */
export abstract class LoggerBase extends Resource {
  public readonly stateful = true;

  /**
   * Logs a message (preflight).
   * @param message The message to log.
   */
  public print(message: string): void {
    // default implementation, can be overridden
    console.log(message);
  }
}

/**
 * A cloud logging facility.
 *
 * @inflight `@winglang/wingsdk.cloud.ILoggerClient`
 */
export class Logger extends LoggerBase {
  /**
   * Returns the logger registered to the given scope, throwing an error if
   * there is none.
   */
  public static of(scope: IConstruct): Logger {
    const logger = (scope as any)[LOGGER_SYMBOL] as Logger;
    if (logger) {
      return logger;
    }

    const parent = scope.node.scope;
    if (!parent) {
      throw new Error("No logger found in scope");
    }

    return Logger.of(parent);
  }

  /**
   * Create a logger and register it to the given scope.
   */
  public static register(scope: IConstruct) {
    const existing = (scope as any)[LOGGER_SYMBOL];
    if (existing !== undefined) {
      throw new Error("There is already a logger registered to this scope.");
    }

    const logger = new Logger(scope, "WingLogger");

    Object.defineProperty(scope, LOGGER_SYMBOL, {
      value: logger,
      enumerable: false,
      writable: false,
    });
  }

  private constructor(scope: Construct, id: string) {
    super(null as any, id);
    return Polycons.newInstance(LOGGER_TYPE, scope, id) as Logger;
  }

  /** @internal */
  public _toInflight(): Code {
    throw new Error("Method not implemented.");
  }
}

/**
 * Inflight interface for `Logger`.
 */
export interface ILoggerClient {
  /**
   * Logs a message. The log will be associated with whichever resource is
   * running the inflight code.
   *
   * NOTICE: this is not an async function because it is wrapped by `console.log()`.
   *
   * @param message The message to print
   * @inflight
   */
  print(message: string): void;
}

/**
 * List of inflight operations available for `Logger`.
 * @internal
 */
export enum LoggerInflightMethods {
  /** `Logger.print` */
  PRINT = "print",
}
