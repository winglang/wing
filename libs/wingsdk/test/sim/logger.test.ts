import * as cloud from "../../src/cloud";
import * as core from "../../src/core";
import { FunctionClient } from "../../src/sim/function.inflight";
import { LoggerClient } from "../../src/sim/logger.inflight";
import * as testing from "../../src/testing";
import { simulatorJsonOf, synthSimulatedApp } from "./util";

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

const INFLIGHT_CODE = core.NodeJsCode.fromInline(`
async function $proc($cap, event) {
  $cap.logger.print("Hello, " + event);
  $cap.logger.print("Wahoo!");
}`);

test("inflight uses a logger", async () => {
  // GIVEN
  const appPath = synthSimulatedApp((scope) => {
    cloud.Logger.register(scope);
    const handler = new core.Inflight({
      code: INFLIGHT_CODE,
      entrypoint: "$proc",
      captures: {
        logger: {
          resource: cloud.Logger.of(scope),
          methods: [cloud.LoggerInflightMethods.PRINT],
        },
      },
    });
    new cloud.Function(scope, "my_function", handler);
  });
  const s = new testing.Simulator({ appPath });
  await s.start();

  const fnAttrs = s.getAttributes("root/my_function");
  const fnClient = new FunctionClient(fnAttrs.functionAddr);

  const loggerAttrs = s.getAttributes("root/WingLogger");
  const loggerClient = new LoggerClient(loggerAttrs.logsDir, "my_function");

  // WHEN
  const PAYLOAD = "Alice";
  await fnClient.invoke(PAYLOAD);

  await sleep(200);

  // THEN
  const logs = await loggerClient.fetchLatestLogs();
  expect(logs).toEqual([
    {
      message: `Hello, ${PAYLOAD}`,
      timestamp: expect.any(Number),
    },
    {
      message: `Wahoo!`,
      timestamp: expect.any(Number),
    },
  ]);
  expect(simulatorJsonOf(appPath)).toMatchSnapshot();

  await s.stop();
});
