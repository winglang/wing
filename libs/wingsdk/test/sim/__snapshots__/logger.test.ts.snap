// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inflight uses a logger 1`] = `
Object {
  "root": Object {
    "children": Object {
      "WingLogger": Object {
        "type": "wingsdk.cloud.Logger",
      },
      "my_function": Object {
        "callees": Array [
          "root/WingLogger",
        ],
        "dependsOn": Array [
          "root/WingLogger",
        ],
        "props": Object {
          "environmentVariables": Object {
            "LOGGER_ADDR__c80f4a947b44da0af65a2258bcf6ee0bfb76f7e65b": "\${root/WingLogger#attrs.logsDir}",
          },
          "sourceCodeFile": "assets/my_function/index.js",
          "sourceCodeLanguage": "javascript",
        },
        "type": "wingsdk.cloud.Function",
      },
    },
    "type": "constructs.Construct",
  },
  "startOrder": Array [
    "root/WingLogger",
    "root/my_function",
  ],
}
`;
